Awful.apk/src/main/java/com/ferg/awfulapp/EmoteFragment.kt

```diff
@@ -0,0 +1,284 @@
+        (parentFragment as EmotePickerListener).onEmoteChosen(emoteCode)
+        EmoteHistory.addRecent(emoteCode)
+        dismiss()
+    }
+}
+
+
+private class EmotePagerAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {
+    data class Page(val fragmentGetter: () -> Fragment, val title: String, val iconResId: Int)
+
+    val pages = arrayOf(
+            Page({ EmoteFragment() }, "Search", R.drawable.ic_search_dark),
+            Page({ EmoteHistoryFragment() }, "Recent", R.drawable.ic_history_dark_24dp)
+    )
+
+    override fun getCount() = pages.size
+    override fun getItem(position: Int) = pages[position].fragmentGetter()
+    // uncomment this to get titles on tabs
+//    override fun getPageTitle(position: Int) = pages[position].title
+}
+
+/**
+ * A fragment that displays all the site emotes, with a live search box to filter them.
+ */
+class EmoteFragment : EmoteGridHaver() {
+
+    override val layoutId = R.layout.emote_picker_main_fragment
+    private val emoteGrid: GridView? by bind(R.id.emote_grid)
+    private val filterText: EditText? by bind(R.id.filter_text)
+
+    override fun getTitle() = "Emotes"
+
+    override fun onActivityCreated(aSavedState: Bundle?) {
+        super.onActivityCreated(aSavedState)
+        emoteGrid?.setUpEmoteGrid()
+
+        // set up the delete button to clear the filter field - which triggers its text change callback
+        bind<ImageButton>(R.id.delete_button).value?.setOnClickListener { filterText!!.setText("") }
+
+        filterText?.apply {
+            setTextColor(ColorProvider.PRIMARY_TEXT.color)
+            addTextChangedListener(object : TextWatcher {
+                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
+                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
+                override fun afterTextChanged(s: Editable) {
+                    emoteLoader.currentFilter = (filterText!!.text.toString().trim())
+                }
+            })
+        }
+    }
+}
+
+
+/**
+ * A fragment that displays the user's most recently used emotes.
+ */
+class EmoteHistoryFragment : EmoteGridHaver() {
+
+    override val layoutId = R.layout.emote_picker_history_fragment
+    private val emoteGrid: GridView? by bind(R.id.emote_grid)
+
+    override fun getTitle() = "Emote history"
+
+    override fun onActivityCreated(aSavedState: Bundle?) {
+        super.onActivityCreated(aSavedState)
+        emoteGrid?.setUpEmoteGrid()
+        emoteLoader.currentFilter = EmoteHistory.getRecent().let { recent ->
+            when {
+                recent.isBlank() -> "match_nothing_thanks" // a placeholder filter that will match nothing... yet
+                else -> recent
+            }
+        }
+    }
+}
+
+
+/**
+ * Class with shared code for the fragment pages, mostly handling the emote grid and lifecycle stuff.
+ */
+abstract class EmoteGridHaver : AwfulFragment() {
```

Comment:
Maybe Owner instead of Haver? Or maybe something else that isn't Haver :V

Reply:
All right, you change-haver!